### ISSUE TITLE
solids4foam crashes on Ubuntu 22-04 (g++ 11.4.0) with OpenFOAM-v2012 and OpenFOAM-v2312.

### ISSUE THREAD

**Issue by philipcardiff** [solids4foam/solids4foam#105]
The beamInCrossflow tutorial crashes within the first few time-steps when running solids4foam on Ubuntu 22-04 (g++ 11.4.0) with OpenFOAM-v2012 and OpenFOAM-v2312.

I suspect the issue comes from the compiler being too new (to be confirmed).

A workaround is to use an older version of Ubuntu with the default older compiler. Maybe using the clang also avoids the problem (as clang on macOS does not have this problem).

**Comment by philipcardiff** [comment]
Fixed in 6c06d8ef.

### Notes
The reason for the behaviour difference between Ubuntu 20.04 and 22.04 is due to the following lines inside `momentumStabilisation.C`:
```
    if (method == "RhieChow")
    {
        result =
        (
           fvc::laplacian(gammaf, vf, "laplacian(DD,D)")
         - fvc::div(gammaf*mesh.Sf() & fvc::interpolate(gradVf)) // This line gives different values!!!
        );
    }
```

Adding an extra set of parentheses makes Ubuntu 20.04 and 22.04 give the same answer:
```
    if (method == "RhieChow")
    {
        result =
        (
           fvc::laplacian(gammaf, vf, "laplacian(DD,D)")
         - fvc::div(gammaf*(mesh.Sf() & fvc::interpolate(gradVf))) // This line now gives the same values
        );
    }
```
At first glance, it seems crazy that such a small difference causes a significant behaviour change; however, on reflection, adding the extra parentheses removes the ambiguity related to the order of multiplication. Before adding the brackets, the compiler has a choice, where it can do one of two things: (i) multiply `gammaf` by `mesh.Sf()` and then multiply by `fvc::interpolate(gradVf)`, or (ii) multiply `mesh.Sf()` by `fvc::interpolate(gradVf)` and then multiply by `gammaf`.
Mathematically, both should give the same answer, but numerically, it seems that the round-off error is causing a difference, i.e. `gammaf` has very large values, while `mesh.Sf()` and `fvc::interpolate(gradVf)` might be very small. It also explains why Clang differs from GCC, as Clang may assume a different multiplication order.
This extra parentheses fixes the issue; nonetheless, it still surprises me.

**Comment by philipcardiff** [comment]
> Fixed in [6c06d8e](https://github.com/solids4foam/solids4foam/commit/6c06d8ef4ff50928d22a21e56c444516db13a8ff).
> 
> ### Notes
> The reason for the behaviour difference between Ubuntu 20.04 and 22.04 is due to the following lines inside `momentumStabilisation.C`:
> 
> ```
>     if (method == "RhieChow")
>     {
>         result =
>         (
>            fvc::laplacian(gammaf, vf, "laplacian(DD,D)")
>          - fvc::div(gammaf*mesh.Sf() & fvc::interpolate(gradVf)) // This line gives different values!!!
>         );
>     }
> ```
> 
> Adding an extra set of parentheses makes Ubuntu 20.04 and 22.04 give the same answer:
> 
> ```
>     if (method == "RhieChow")
>     {
>         result =
>         (
>            fvc::laplacian(gammaf, vf, "laplacian(DD,D)")
>          - fvc::div(gammaf*(mesh.Sf() & fvc::interpolate(gradVf))) // This line now gives the same values
>         );
>     }
> ```
> 
> At first glance, it seems crazy that such a small difference causes a significant behaviour change; however, on reflection, adding the extra parentheses removes the ambiguity related to the order of multiplication. Before adding the brackets, the compiler has a choice, where it can do one of two things: (i) multiply `gammaf` by `mesh.Sf()` and then multiply by `fvc::interpolate(gradVf)`, or (ii) multiply `mesh.Sf()` by `fvc::interpolate(gradVf)` and then multiply by `gammaf`. Mathematically, both should give the same answer, but numerically, it seems that the round-off error is causing a difference, i.e. `gammaf` has very large values, while `mesh.Sf()` and `fvc::interpolate(gradVf)` might be very small. It also explains why Clang differs from GCC, as Clang may assume a different multiplication order. This extra parentheses fixes the issue; nonetheless, it still surprises me.

@iBatistic : you might find this interesting (and scary).

**Comment by iBatistic** [comment]
@philipcardiff, totally crazy. It is impressive that you managed to figure this out.
